pyteal/ast/subroutine.py

```diff
@@ -610,32 +611,71 @@ def method_signature(self, overriding_name: str = None) -> str:
-        skip_names = ["return", "output"]
-
-        args = [
-            {
-                "type": str(abi.type_spec_from_annotation(val)),
-                "name": name,
-            }
-            for name, val in self.subroutine.annotations.items()
-            if name not in skip_names
-        ]
-
-        spec = {
-            "name": self.name(),
-            "args": args,
-            "returns": {"type": str(self.type_of())},
-        }
+        desc: str = ""
+        arg_descs: dict[str, str] = {}
+        return_desc: str = ""
+        args: list[dict[str, str]] = []
+
+        if self.subroutine.implementation.__doc__:
+            docstring = parse_docstring(self.subroutine.implementation.__doc__)
+
+            # Combine short and long descriptions with newline
+            method_desc = (
+                (docstring.short_description if docstring.short_description else "")
+                + "\n"
+                + (docstring.long_description if docstring.long_description else "")
+            )
 
-        if self.subroutine.implementation.__doc__ is not None:
-            spec["desc"] = " ".join(
+            # Turn double new line into single, replacing single newline with space
+            desc = "\n".join(
                 [
-                    i.strip()
-                    for i in self.subroutine.implementation.__doc__.split("\n")
+                    i.replace("\n", " ").strip()
+                    for i in method_desc.split("\n\n")
                     if not (i.isspace() or len(i) == 0)
                 ]
             )
 
+            # Get the descriptions for any documented arguments
+            arg_descs = {
+                arg.arg_name: arg.description
+                for arg in docstring.params
+                if arg.arg_name != self.OUTPUT_ARG_NAME and arg.description is not None
+            }
+
+            # Get the special return description
+            return_desc = (
+                ""
+                if not docstring.returns or not docstring.returns.description
+                else docstring.returns.description
+            )
+
+            # Generate the ABI method object given the subroutine args
+            # Add in description if one is set
+            for name, val in self.subroutine.annotations.items():
+                # Skip annotations for `return` and `output` in the args list
+                if name in ["return", self.OUTPUT_ARG_NAME]:
+                    continue
+
+                arg_obj = {
+                    "type": str(abi.type_spec_from_annotation(val)),
+                    "name": name,
+                }
+
+                if name in arg_descs:
+                    arg_obj["desc"] = arg_descs[name]
+
+                args.append(arg_obj)
+
+        # Create the return obj for the method, adding description if set
+        return_obj = {"type": str(self.type_of())}
+        if return_desc:
+            return_obj["desc"] = return_desc
```

Comment:
Suggestion: don't add this description if the method returns void.

Reply:
I did this, but I'm not sure if we should skip it. 

If the author takes the time to write a return description, why not add it?

Reply:
The arguments as I see it are:

### In favor of always annotating the return description:
* If the author wrote a return annotation, might as well carry it forward to the contract JSON. They can always proofread the generated JSON and remove the annotation if they see fit.

### Against always annotating the return description:
* For an ABI method that returns void, there is no reason to write a return description; anything you have to say about what the function does should go into the general function description. Yes, PyTeal method implementations *do* return an `Expr`, but this is only a language-level convention and irrelevant to an external caller.
* I suspect (but don't know how common this really is) that developers tend to write return annotations in the form of "Returns a PyTeal expression that..." (that's at least what we do internally). This type of explanation is meaningless outside of the realm of PyTeal and would only serve to add noise to a contract JSON.

I'm still slightly in favor of the current behavior (dropping return description if return is void), but I don't feel so strongly about it.